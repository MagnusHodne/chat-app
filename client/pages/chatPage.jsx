import React, { useContext, useEffect, useState } from "react";
import { useLoading } from "../lib/useLoading";
import {
  ErrorComponent,
  LoadingComponent,
} from "../components/feedbackComponents";
import { ChatApiContext } from "../chatApiContext";
import { ChatComponent } from "../components/chatComponents";

function ChatConnection({ initialMessages, user }) {
  const [messages, setMessages] = useState(initialMessages);
  const [ws, setWs] = useState({});

  //Establishes a websocket on component mount
  useEffect(() => {
    const protocol = window.location.protocol;
    const ws = new WebSocket(
      (protocol === "http:" ? "ws://" : "wss://") + window.location.host
    );
    ws.onopen = (event) => {
      console.log("Opened", event);
    };
    ws.onmessage = (event) => {
      const res = JSON.parse(event.data);
      if (res.action === "create") {
        const { message, user, created, _id } = res;
        setMessages((oldMessages) => [
          ...oldMessages,
          { message, user, created, _id },
        ]);
      }
      if (res.action === "delete") {
        const { _id } = res;
        setMessages((oldMessages) =>
          oldMessages.filter((msg) => msg._id !== _id)
        );
      }
    };
    setWs(ws);
  }, []);

  function handleNewMessage(message) {
    ws.send(JSON.stringify({ action: "create", message, user }));
  }
  function handleDeleteMessage(_id) {
    ws.send(JSON.stringify({ action: "delete", _id }));
  }

  return (
    <ChatComponent
      messages={messages}
      onNewMessage={handleNewMessage}
      onDeleteMessage={handleDeleteMessage}
    />
  );
}

export function ChatPage({ user }) {
  const { fetchChatLog } = useContext(ChatApiContext);
  const { data, loading, error } = useLoading(fetchChatLog);
  if (loading)
    return <LoadingComponent message={"Fetching chat log, please wait..."} />;
  if (error)
    return (
      <ErrorComponent error={"Unable to fetch messages, please try again"} />
    );
  return <ChatConnection user={user} initialMessages={data} />;
}
