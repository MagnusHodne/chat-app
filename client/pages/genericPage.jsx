import React from "react";
import { Content, Header, Sidebar } from "../components/layoutComponents";

export function GenericPage({ sidebarContent, mainContent }) {
  return (
    <div className={"page"}>
      <Header />
      <Sidebar content={sidebarContent} />
      <Content content={mainContent} />
    </div>
  );
}
