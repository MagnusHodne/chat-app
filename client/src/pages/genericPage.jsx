import React from "react";
import { Content, Header, Sidebar } from "../components/layoutComponents";

export function GenericPage({ sidebarContent, mainContent }) {
  return (
    <div
      className={
        "grid h-[95vh] grid-cols-[25%_75%] grid-rows-[5em_1fr] gap-0 overflow-clip rounded-3xl p-2"
      }
    >
      <Header />
      <Sidebar content={sidebarContent} />
      <Content content={mainContent} />
    </div>
  );
}
