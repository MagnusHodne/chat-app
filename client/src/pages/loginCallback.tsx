import React, { useEffect } from "react";
import {
  ErrorComponent,
  LoadingComponent,
} from "../components/feedbackComponents";
import { useAuth0 } from "@auth0/auth0-react";
import { useNavigate } from "react-router-dom";

type CallbackProps = {
  setUserInfo: (userInfo: any) => void;
};

export const LoginCallback: React.FC<CallbackProps> = ({ setUserInfo }) => {
  const { error, user, getAccessTokenSilently } = useAuth0();
  const navigate = useNavigate();

  useEffect(() => {
    const processLogin = async () => {
      const accessToken = await getAccessTokenSilently();
      localStorage.removeItem("accessToken");
      localStorage.setItem("accessToken", accessToken);
      localStorage.removeItem("user");
      localStorage.setItem("user", JSON.stringify(user));
      setUserInfo(user);

      //TODO - make call to new user endpoint here, let backend handle registration if new user
      //await CallBackendApi("user/new", "post", user)

      navigate("/home");
    };
    processLogin();
  });

  if (error) {
    return (
      <ErrorComponent
        error={`Failed to redirect from login: ${error.message}`}
      />
    );
  }
  return <LoadingComponent message={"Redirecting, please wait..."} />;
};
