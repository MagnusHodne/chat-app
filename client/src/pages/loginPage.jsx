import React, { useContext, useEffect, useState } from "react";
import { Route, Routes, useNavigate, useParams } from "react-router-dom";
import { useLoading } from "../lib/useLoading";
import {
  ErrorComponent,
  LoadingComponent,
} from "../components/feedbackComponents";
import { ChatApiContext } from "../chatApiContext";

/**
 * Function that displays a waiting message to the user, posts the access_token to the server
 * (whichs signs it), then returns to the front page
 */
export function LoginCallback({ reload, config }) {
  const { provider } = useParams();
  const [error, setError] = useState("");
  const navigate = useNavigate();
  const urlAfterLogin = "/chat";
  const { registerLogin, getCallbackParameters, getExpectedState } =
    useContext(ChatApiContext);

  useEffect(async () => {
    const { access_token, error, error_description, state, code } =
      getCallbackParameters();

    const expected_state = getExpectedState();
    if (!state || expected_state !== state) {
      setError("Unexpected state");
      return;
    }

    if (error || error_description) {
      setError(`Error: ${error} (${error_description})`);
      return;
    }

    if (code) {
      const { client_id, token_endpoint } = config[provider];
      const code_verifier = window.sessionStorage.getItem("code_verifier");
      const payload = {
        grant_type: "authorization_code",
        code,
        client_id,
        code_verifier,
        redirect_uri: `${window.location.origin}/login/${provider}/callback`,
      };
      const res = await fetch(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      if (!res.ok) {
        setError(`Failed to fetch token ${res.status}: ${await res.text()}`);
        return;
      }
      const { access_token } = await res.json();
      await registerLogin(provider, { access_token });

      //This updates the user info for the site if we logged in with a code flow
      reload();
      navigate(urlAfterLogin);
      return;
    }

    if (!access_token) {
      setError("Missing access_token");
      return;
    }

    await registerLogin(provider, { access_token });
    //This updates the user info for the site if we logged in with a token flow
    reload();
    navigate(urlAfterLogin);
  }, []);

  if (error) {
    return <ErrorComponent error={error} />;
  }

  return <LoadingComponent message={"Fetching user data, please wait..."} />;
}

export function Login() {
  const { handleLogin, fetchConfig } = useContext(ChatApiContext);
  const { error, data, loading } = useLoading(() => fetchConfig());

  //This react component actually handles the login. Have to nest it in order to fetch config first...
  const LoginHandler = () => {
    const { provider } = useParams();
    const { error } = useLoading(() => handleLogin(provider, data));
    if (error) {
      return <ErrorComponent error={error} />;
    }

    return <LoadingComponent message={"Redirecting to login, please wait"} />;
  };
  if (error) {
    return <ErrorComponent error={error} />;
  }
  if (loading) {
    return <LoadingComponent message={"Fetching config"} />;
  }

  return <LoginHandler />;
}

export function Logout({ reload }) {
  const navigate = useNavigate();
  const { endSession } = useContext(ChatApiContext);
  useEffect(async () => {
    await endSession();
    reload();
    navigate("/");
  });
  return <LoadingComponent message={"Logging out, please wait..."} />;
}

export function LoginPage({ config, reload }) {
  return (
    <Routes>
      <Route path={"/:provider"} element={<Login />} />
      <Route
        path={"/:provider/callback"}
        element={<LoginCallback config={config} reload={reload} />}
      />
      <Route path={"/endsession"} element={<Logout reload={reload} />} />
    </Routes>
  );
}
