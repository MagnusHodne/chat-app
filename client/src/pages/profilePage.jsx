import React, { useContext, useState } from "react";
import { Button } from "../components/basics";
import { useLoading } from "../lib/useLoading";
import { ChatApiContext } from "../chatApiContext";
import {
  ErrorComponent,
  LoadingComponent,
} from "../components/feedbackComponents";
import { UserContext } from "../userContext";
import { Avatar } from "../components/avatar";
import { PaddedContent } from "../components/layoutComponents";

export function ProfilePage() {
  const { fetchUserInfo } = useContext(ChatApiContext);
  const { user } = useContext(UserContext);
  const { data, loading, error } = useLoading(() =>
    fetchUserInfo({ sub: user.sub })
  );
  if (loading)
    return <LoadingComponent message={"Fetching user bio, please wait..."} />;
  if (error) return <ErrorComponent error={"Unable to fetch user bio"} />;
  return <ProfileCard userinfo={data} initBio={data.bio} />;
}

function ProfileCard({ userinfo, initBio }) {
  const { updateUserBio } = useContext(ChatApiContext);

  const [bio, setBio] = useState(initBio);
  async function handleSubmit(e) {
    e.preventDefault();
    await updateUserBio({ bio, sub: userinfo.sub });
  }

  return (
    <PaddedContent
      content={
        <>
          <div className={"h-24 w-24"}>
            <Avatar src={userinfo.picture} />
          </div>
          <h1 className={"text-2xl"}>{`${userinfo.name}`}</h1>
          <form onSubmit={handleSubmit}>
            <label>
              <strong>Bio</strong>
              <textarea
                className={
                  "mx-0 my-1 w-full resize-none overflow-y-auto rounded-md border-none bg-thischord-500 pt-2 pl-2"
                }
                name="text"
                value={bio}
                rows={10}
                onChange={(e) => setBio(e.target.value)}
              ></textarea>
            </label>
            <Button title={"Update bio"} />
          </form>
        </>
      }
    />
  );
}
