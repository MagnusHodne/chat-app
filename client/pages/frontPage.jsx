import React from "react";
import { Route, Routes } from "react-router-dom";
import { GenericPage } from "./genericPage";
import { ProfilePage } from "./profilePage";
import { ChatPage } from "./chatPage";
import { NotFoundPage } from "./notFoundPage";
import { Button } from "../components/basics";
import { AuthorizedComponent } from "../components/authorizedComponent";

export function FrontPage({ user }) {
  function ContentRouter({ user }) {
    return (
      <Routes>
        <Route path={"/profile"} element={<ProfilePage user={user} />} />
        <Route path={"/chat"} element={<ChatPage user={user} />} />
        <Route path={"/*"} element={<NotFoundPage />} />
      </Routes>
    );
  }

  function SidebarContent() {
    return (
      <div style={{ display: "flex", flexDirection: "column" }}>
        <Button title={"Profile"} icon={"fa-solid fa-user"} to={"/profile"} />
        <Button title={"Chat"} icon={"fa-solid fa-comment"} to={"/chat"} />
      </div>
    );
  }
  return (
    <AuthorizedComponent
      user={user}
      successComponent={
        <GenericPage
          user={user}
          mainContent={<ContentRouter user={user} />}
          sidebarContent={<SidebarContent user={user} />}
        ></GenericPage>
      }
    />
  );
}
