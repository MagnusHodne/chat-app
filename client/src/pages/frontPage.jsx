import React from "react";
import { Route, Routes } from "react-router-dom";
import { GenericPage } from "./genericPage";
import { ProfilePage } from "./profilePage";
import { ChatPage } from "./chatPage";
import { NotFoundPage } from "./notFoundPage";
import { Button } from "../components/basics";
import { AuthorizedComponent } from "../components/authorizedComponent";

export function FrontPage() {
  function ContentRouter() {
    return (
      <Routes>
        <Route path={"/profile"} element={<ProfilePage />} />
        <Route path={"/chat"} element={<ChatPage />} />
        <Route path={"/*"} element={<NotFoundPage />} />
      </Routes>
    );
  }

  function SidebarContent() {
    return (
      <div style={{ display: "flex", flexDirection: "column" }}>
        <Button title={"Profile"} icon={"fa-solid fa-user"} to={"/profile"} />
        <Button title={"Chat"} icon={"fa-solid fa-comment"} to={"/chat"} />
      </div>
    );
  }
  return (
    <AuthorizedComponent
      successComponent={
        <GenericPage
          mainContent={<ContentRouter />}
          sidebarContent={<SidebarContent />}
        ></GenericPage>
      }
    />
  );
}
