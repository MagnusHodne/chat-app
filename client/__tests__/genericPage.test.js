import React from "react";
import ReactDOM from "react-dom";
import { GenericPage } from "../src/pages/genericPage";
import { MemoryRouter } from "react-router-dom";

describe("GenericPage", () => {
  it("matches snapshot", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <GenericPage />
      </MemoryRouter>,
      element
    );
    expect(element).toMatchSnapshot();
  });
});
