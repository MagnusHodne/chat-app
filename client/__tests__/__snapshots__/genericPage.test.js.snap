// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericPage matches snapshot 1`] = `
<div>
  <div
    class="m-2 grid h-full grid-cols-[25%_75%] grid-rows-[5em_1fr] gap-0 overflow-clip rounded p-0"
  >
    <header
      class="col-span-2 flex flex-row items-center justify-between border-b-2 border-solid border-thischord-900 bg-thischord-600 p-4"
    >
      <div
        class="flex flex-row items-center gap-3"
      >
        <i
          class="fa-solid fa-gamepad"
        />
        <h1
          class="text-2xl font-black"
        >
          Thischord
        </h1>
      </div>
      <div />
    </header>
    <div
      class="bg-thischord-800 p-2"
    />
    <div
      class="min-h-0 bg-thischord-600"
    />
  </div>
</div>
`;
